runtime/hydradx/src/evm/erc20_currency.rs

```diff
@@ -0,0 +1,267 @@
+		// to
+		data.extend_from_slice(H256::from(to).as_bytes());
+		// amount
+		data.extend_from_slice(H256::from_uint(&U256::from(value.saturated_into::<u128>())).as_bytes());
+
+		handle_result(Executor::<T>::call(context, data, U256::zero(), 200_000))
+	}
+
+	fn transfer_from(context: CallContext, from: EvmAddress, to: EvmAddress, value: Balance) -> DispatchResult {
+		let mut data = Into::<u32>::into(Function::TransferFrom).to_be_bytes().to_vec();
+		// from
+		data.extend_from_slice(H256::from(from).as_bytes());
+		// to
+		data.extend_from_slice(H256::from(to).as_bytes());
+		// amount
+		data.extend_from_slice(H256::from_uint(&U256::from(value.saturated_into::<u128>())).as_bytes());
+
+		handle_result(Executor::<T>::call(context, data, U256::zero(), 200_000))
+	}
+}
+
+fn decode_string(output: Vec<u8>) -> Option<Vec<u8>> {
+	// output is 32-byte aligned and consists of 3 parts:
+	// - part 1: 32 byte, the offset of its description is passed in the position of
+	// the corresponding parameter or return value.
+	// - part 2: 32 byte, string length
+	// - part 3: string data
+	if output.len() < 64 || output.len() % 32 != 0 {
+		return None;
+	}
+
+	let offset = U256::from_big_endian(&output[0..32]);
+	let length = U256::from_big_endian(&output[offset.as_usize()..offset.as_usize() + 32]);
+	if output.len() < offset.as_usize() + 32 + length.as_usize() {
+		return None;
+	}
+
+	let mut data = Vec::new();
+	data.extend_from_slice(&output[offset.as_usize() + 32..offset.as_usize() + 32 + length.as_usize()]);
+
+	Some(data.to_vec())
+}
+
+fn decode_integer(value: Vec<u8>) -> Option<U256> {
+	if value.len() != 32 {
+		return None;
+	}
+	U256::checked_from(value.as_slice())
+}
+
+fn handle_result(result: CallResult) -> DispatchResult {
+	let (exit_reason, value) = result;
+	match exit_reason {
+		ExitReason::Succeed(ExitSucceed::Returned) => Ok(()),
+		ExitReason::Succeed(ExitSucceed::Stopped) => Ok(()),
+		_ => Err(DispatchError::Other(&*Box::leak(
+			format!("evm:0x{}", hex::encode(value)).into_boxed_str(),
+		))),
+	}
+}
+
+impl<T> MultiCurrency<T::AccountId> for Erc20Currency<T>
+where
+	T: Config + pallet_evm::Config,
+	pallet_evm_accounts::Pallet<T>: InspectEvmAccounts<T::AccountId>,
+	T::AccountId: AsRef<[u8; 32]> + IsType<AccountId32>,
+	BalanceOf<T>: TryFrom<U256> + Into<U256>,
+{
+	type CurrencyId = EvmAddress;
+	type Balance = Balance;
+
+	fn minimum_balance(_contract: Self::CurrencyId) -> Self::Balance {
+		Zero::zero()
+	}
+
+	fn total_issuance(contract: Self::CurrencyId) -> Self::Balance {
+		<Self as ERC20>::total_supply(CallContext {
+			contract,
+			sender: Default::default(),
+			origin: Default::default(),
+		})
+	}
+
+	fn total_balance(contract: Self::CurrencyId, who: &T::AccountId) -> Self::Balance {
+		Self::free_balance(contract, who)
+	}
+
+	fn free_balance(contract: Self::CurrencyId, who: &T::AccountId) -> Self::Balance {
+		<Self as ERC20>::balance_of(CallContext::new_view(contract), EvmAccounts::<T>::evm_address(who))
```

Comment:
What if the asset is non-transferrable? (i.e. debt token)

Reply:
idk, depends on implementation of the erc20, i am just mapping `balance_of` here, i cannot ensure that it will be really transferable, that always depends on erc20.transfer implementation